Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,2354 @@
+}
+
+//------------------------------------------------------------------------------
+
+const Mobod& CableSpan::Impl::getOriginBody() const
+{
+    return getSubsystem()
+        .getImpl()
+        .getMultibodySystem()
+        .getMatterSubsystem()
+        .getMobilizedBody(m_OriginBody);
+}
+
+const Mobod& CableSpan::Impl::getTerminationBody() const
+{
+    return getSubsystem()
+        .getImpl()
+        .getMultibodySystem()
+        .getMatterSubsystem()
+        .getMobilizedBody(m_TerminationBody);
+}
+
+//------------------------------------------------------------------------------
+
+size_t CableSpan::Impl::countActive(const State& s) const
+{
+    size_t count = 0;
+    for (const CurveSegment& segment : m_CurveSegments) {
+        if (segment.getInstanceEntry(s).isInContactWithSurface()) {
+            ++count;
+        }
+    }
+    return count;
+}
+
+int CableSpan::Impl::calcPathPoints(
+    const State& state,
+    Real maxLengthIncrement,
+    std::function<void(Real length, Vec3 point_G)> sink) const
+{
+    // Count number of points written.
+    int count = 0;
+
+    // Write the initial point.
+    const PosInfo& pos = getPosInfo(state);
+    sink(0., pos.originPoint_G);
+    ++count;
+
+    // Write points along each of the curves.
+    for (const CurveSegment& curve : m_CurveSegments) {
+
+        Real curveLength = curve.getInstanceEntry(state).length;
+        int nPoints =
+            static_cast<int>(std::abs(curveLength / maxLengthIncrement) + 1.);
+        nPoints = nPoints < 2 ? 2 : nPoints;
+
+        count += curve.calcPathPoints(state, sink, nPoints);
+    }
+
+    // Write the termination point.
+    sink(pos.cableLength, pos.terminationPoint_G);
+    ++count;
+
+    // Return number of points written.
+    return count;
+}
+
+//------------------------------------------------------------------------------
+
+const CurveSegment* CableSpan::Impl::findPrevActiveCurveSegment(
+    const State& s,
+    ObstacleIndex ix) const
+{
+    for (int i = ix - 1; i >= 0; --i) {
+        // Find the active segment before the current.
+        if (m_CurveSegments.at(ObstacleIndex(i))
+                .getInstanceEntry(s)
+                .isInContactWithSurface()) {
+            return &m_CurveSegments.at(ObstacleIndex(i));
+        }
+    }
+    return nullptr;
+}
+
+const CurveSegment* CableSpan::Impl::findNextActiveCurveSegment(
+    const State& s,
+    ObstacleIndex ix) const
+{
+    // Find the active segment after the current.
+    for (int i = ix + 1; i < m_CurveSegments.size(); ++i) {
+        if (m_CurveSegments.at(ObstacleIndex(i))
+                .getInstanceEntry(s)
+                .isInContactWithSurface()) {
+            return &m_CurveSegments.at(ObstacleIndex(i));
+        }
+    }
+    return nullptr;
+}
+
+Vec3 CableSpan::Impl::findPrevPoint(const State& s, ObstacleIndex ix) const
+{
+    auto CalcCurveFinalContactPoint = [&](const CurveSegment& curve) -> Vec3 {
```

Comment:
Function definitions (incl. lambas) usually get a newline for the opening bracer